### PR TITLE
drivers: spi: Add shim for nrfx SPIS driver

### PR DESCRIPTION
**Pull request by anangl**
This adds a translation layer to make the nrfx driver for the nRF SPIS
(SPI Slave with EasyDMA) peripheral accessible via the Zephyr's API.

Depends on #6985, in particular on the commit bb66f3f3704ed3adcda7e46c8125fed2c54af91f.

### CODE REVIEW

File: ext/hal/nordic/nrfx_glue.h [hunk 1]
```diff
@@ -176,6 +176,37 @@ extern "C" {
...
+ * the peripheral base address, i.e. the address of a fixed block of 0x1000
+ * bytes of address space assigned to this peripheral.
+ * See the chapter "Peripheral interface" (sections "Peripheral ID" and
+ * "Interrupts") in the product specification of a given SoC.
+ *
+ * @param[in] base_addr  Peripheral base address.
+ *
+ * @return Interrupt number associated with the specified peripheral.
+ */
+/* TODO - this macro should become a part of <nrfx_common.h> */
```
Comment (carlescufi) [inline ext/hal/nordic/nrfx_glue.h]: Can you add an entry in the nrfx bug tracking system to get rid of this once it's part of nrfx_common.h ?

**Comment by anangl** [comment]
@tbursztyka I have some doubts regarding `spi_context`:
1. In slave mode it does not automatically release the lock, you need to explicitly call `spi_release` before calling `spi_transceive` again. Is it still intented? The expected behavior is not documented. IIRC it was earlier, but it is no longer after the last update.
2. It returns properly the number of received frames (`ctx->recv_frames`) only in a synchronous operation. In an asynchronous one this value should be passed as the parameter to `k_poll_signal`, because when `spi_transceive_async` returns, it is not known yet.
3. When you provide both master and slave implementations, `ctx->recv_frames` is counted in both modes (even if they are done in separate code, like in the nRF case), and not used then for master. Wouldn't it be better to leave it to particular slave implementations to provide this value in the most convenient way for them? As you can see [here](https://github.com/zephyrproject-rtos/zephyr/pull/7003/commits/cd297fbe6788d1b20f4f209832fc5885c2ab0a9e#diff-4550e1ad9023ccde2007205a5a77b10bR225), in nRF implementation I can get it from a peripheral register after the transfer is done.

File: ext/hal/nordic/nrfx_glue.h (hunk 1 above)
Comment (anangl) [reply ext/hal/nordic/nrfx_glue.h]: [Done](https://github.com/NordicSemiconductor/nrfx/issues/18).

File: drivers/spi/spi_nrfx_spi.c [hunk 2]
```diff
@@ -270,24 +270,12 @@ static int init_spi(struct device *dev, const nrfx_spi_config_t *config)
...
- * this peripheral. See the chapter "Peripheral interface" (sections "Peripheral
- * ID" and "Interrupts") in the product specification of a given SoC.
- */
-#define NRF_SPI_IRQ_NUMBER(idx)  (uint8_t)((uint32_t)NRF_SPI##idx >> 12u)
-
 #define SPI_NRFX_SPI_DEVICE(idx)					\
 	static int spi_##idx##_init(struct device *dev)			\
 	{								\
-		IRQ_CONNECT(NRF_SPI_IRQ_NUMBER(idx),			\
+		IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_SPI##idx),		\
```
Comment (galak) [inline drivers/spi/spi_nrfx_spi.c]: Why isn't the IRQ number coming from device tree?

**Comment by tbursztyka** [comment]
@anangl 
1) That was the intention indeed, but it's a left-over of the first slave attemps. will get rid of that behavior.
1) yep that's was forgotten
1) why not, but that was just super simple to do it in spi_context.h (and actually, it makes things easy for the previous point)

Will send patches

**Review by tbursztyka — APPROVED** [review]

**Comment by nashif** [comment]
is this still DNM?

**Comment by anangl** [comment]
@nashif Now that #6985 is merged, this one is no longer a DNM.

**Comment by anangl** [comment]
Added one more commit that updates the default configuration for 96b_carbon_nrf51.

**Comment by carlescufi** [comment]
@anangl there's still an unanswered question from @galak about why the IRQ number is not coming from DT. I assume  you will send a PR to switch SPI to DT later?

**Comment by tbursztyka** [comment]
@carlescufi I have one PR being prepared for moving all spi (controllers & devices) to DT.

**Review by carlescufi — APPROVED** [review]

**Comment by anangl** [comment]
@carlescufi Yes, that was the plan. And I answered the same question from @galak in #6985. I forgot about the one here.
But now I'll wait for the PR from @tbursztyka.

**Comment by carlescufi** [comment]
@anangl ok, so if you fix CI we can merge this then

**Comment by anangl** [comment]
Added a commit that made CI green. Ready to merge, I guess.

**Pull request merged**